torch/distributed/fsdp/_wrap_utils.py

```diff
@@ -119,3 +120,98 @@ def _warn_on_overridden_mixed_precision(
         "These modules will be wrapped as separate FSDP instacnes with mixed "
         "precision disabled."
     )
+
+
+def _validate_frozen_params(
+    root_module: nn.Module,
+    modules_to_wrap: Set[nn.Module],
+    ignored_params: Set[nn.Parameter],
+    strict: bool,
+):
+    """
+    This checks that, given ``modules_to_wrap``, each module would manage
+    parameters that are uniformly frozen or non-frozen. This uniformity
+    requirement is strict for ``use_orig_params=False`` (hard error) and highly
+    recommended for ``use_orig_params=True`` (user warning).
+    """
+    visited_modules = {root_module}
+    stack = [("", root_module)]
+    topo_sorted_named_modules: List[Tuple[str, nn.Module]] = []
+    while stack:
+        module_name, module = stack.pop()
+        topo_sorted_named_modules.append((module_name, module))
+        for child_module_name, child_module in module.named_children():
+            if child_module is None:  # only for overrides of `named_children()`
+                continue
+            if child_module not in visited_modules:
+                visited_modules.add(child_module)
+                if module_name != "":
+                    child_module_name = module_name + "." + child_module_name
+                stack.append((child_module_name, child_module))
+    reverse_topo_sorted_modules = reversed(topo_sorted_named_modules)
+    visited_modules.clear()
+    for module_name, module in reverse_topo_sorted_modules:
+        if module in modules_to_wrap:
+            param_to_fqn = _get_param_to_fqn(
+                module, ignored_params, visited_modules, module_name
+            )
+            frozen_param_fqns: List[str] = []
+            nonfrozen_param_fqns: List[str] = []
+            for param, fqn in param_to_fqn.items():
+                if param.requires_grad:
+                    nonfrozen_param_fqns.append(fqn)
+                else:
+                    frozen_param_fqns.append(fqn)
+            if len(frozen_param_fqns) > 0 and len(nonfrozen_param_fqns) > 0:
+                msg = f"{module_name} has both parameters with requires_grad=True and False."
+                if strict:
+                    msg += " FSDP does not support wrapping such modules.\n"
+                else:
+                    msg += (
+                        " FSDP does not recommend wrapping such modules since "
+                        "the gradient memory usage will be higher than expected.\n"
+                    )
+                msg += (
+                    f"The following parameters have requires_grad=True:\n{nonfrozen_param_fqns}\n"
+                    f"The following parameters have requires_grad=False:\n{frozen_param_fqns}"
+                )
+                if strict:
+                    raise ValueError(msg)
+                else:
+                    warnings.warn(msg)
```

Comment:
add a line saying the recommended workaround - wrap frozen parameters into their own FSDP unit?